### PR TITLE
Split Treehouse lifecycle out of RedwoodView

### PR DESCRIPTION
**Pull request by swankjesse**
https://github.com/cashapp/redwood/issues/2392

---

- [x] `CHANGELOG.md`'s "Unreleased" section has been updated, if applicable.

### CODE REVIEW

**Comment by swankjesse** [comment]
My BIG INSIGHT in this PR is that everything works better if I expose content state two ways:
 - First as child components. So `TreehouseViews` can show loading spinners and sad-tab icons inline if that’s their thing.
 - ALSO as state. So `TreehouseViews` can also display that stuff outside of the content if that’s their thing.

All of my previous attempts were clumsy because they treated these two cases as symmetric.

File: redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/HostProtocolAdapter.kt [hunk 1]
```diff
@@ -425,7 +425,7 @@ private class RootProtocolNode<W : Any>(
     }
 
   override fun detach() {
-    children.detach()
+    // Do nothing because 'children' is owned by the host's TreehouseView.
```
Comment (JakeWharton) [inline redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/HostProtocolAdapter.kt]: The Redwood protocol doesn't _strictly_ have to be used by Treehouse, although that is certainly the overwhelmingly common usage. Maybe put a note on the class docs up above, or on `close()`.

**Review by JakeWharton — APPROVED** [review]

File: redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/HostProtocolAdapter.kt (hunk 1 above)
Comment (swankjesse) [reply redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/HostProtocolAdapter.kt]: Yeah, good point. I should have said `RedwoodView` here instead.

Comment (JakeWharton) [reply redwood-protocol-host/src/commonMain/kotlin/app/cash/redwood/protocol/host/HostProtocolAdapter.kt]: Oh even better. I thought it was something specific to treehouse.

**Pull request merged**